internal/state/config/config.go

```diff
@@ -1,178 +1,248 @@
+	cfg.v.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
+
+	cfg.fs = pflag.NewFlagSet("hcloud", pflag.ContinueOnError)
+	for _, o := range Options {
+		o.addToFlagSet(cfg.fs)
+	}
+	if err := cfg.v.BindPFlags(cfg.fs); err != nil {
+		panic(err)
+	}
+}
+
+// ReadConfig reads the config from the flags, env and the given config file f.
+// See [ParseConfigFile] for the supported types of f.
+func ReadConfig(cfg Config, f any) error {
+
+	// error is ignored since invalid flags are already handled by cobra
+	_ = cfg.FlagSet().Parse(os.Args[1:])
+
+	// load env already so we can determine the active context
+	cfg.Viper().AutomaticEnv()
+
+	return cfg.ParseConfigFile(f)
 }
 
-func ReadConfig(path string) (Config, error) {
-	cfg := &config{path: path}
+// ParseConfigFile parses the given config file f.
+// f can be of the following types:
+// - nil: the default config file is used
+// - string: the path to the config file
+// - io.Reader: the config is read from the reader
+// - []byte: the config is read from the byte slice
+// - any other type: an error is returned
+func (cfg *config) ParseConfigFile(f any) error {
+	var (
+		cfgBytes []byte
+		err      error
+	)
+
+	cfg.path = OptionConfig.Get(cfg)
+	path, ok := f.(string)
+	if path != "" && ok {
+		cfg.path = path
+	}
 
-	_, err := os.Stat(path)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return cfg, nil
+	if f == nil || ok {
+		// read config from file
+		cfgBytes, err = os.ReadFile(cfg.path)
+		if err != nil {
+			return err
+		}
+	} else {
+		switch f := f.(type) {
+		case io.Reader:
+			cfgBytes, err = io.ReadAll(f)
+			if err != nil {
+				return err
+			}
+		case []byte:
+			cfgBytes = f
+		default:
+			return fmt.Errorf("invalid config file type %T", f)
 		}
-		return cfg, err
 	}
 
-	data, err := os.ReadFile(path)
-	if err != nil {
-		return nil, err
+	if err := toml.Unmarshal(cfgBytes, &cfg.schema); err != nil {
+		return err
 	}
 
-	if err = cfg.unmarshal(data); err != nil {
-		return nil, err
+	if cfg.schema.ActiveContext != "" {
+		// ReadConfig resets the current config and reads the new values
+		// We don't use viper.Set here because of the value hierarchy. We want the env and flags to
+		// be able to override the currently active context. viper.Set would take precedence over
+		// env and flags.
+		err = cfg.v.ReadConfig(bytes.NewReader([]byte(fmt.Sprintf("context = %q\n", cfg.schema.ActiveContext))))
+		if err != nil {
+			return err
+		}
 	}
 
-	return cfg, nil
-}
+	// read active context from viper
+	activeContext := cfg.schema.ActiveContext
+	if ctx := OptionContext.Get(cfg); ctx != "" {
+		activeContext = ctx
+	}
 
-func (cfg *config) Write() error {
-	data, err := cfg.marshal()
-	if err != nil {
-		return err
+	cfg.contexts = cfg.schema.Contexts
+	for i, ctx := range cfg.contexts {
+		if ctx.ContextName == activeContext {
+			cfg.activeContext = cfg.contexts[i]
+			break
+		}
 	}
-	if err := os.MkdirAll(filepath.Dir(cfg.path), 0777); err != nil {
-		return err
```

Comment:
Happy not to see this 0777 in the new code, but shouldn't we keep the config dir creation behavior?